README.md
# Music_player
online music player 


history.txt
Heading Home	[World Of Walker]
Jonas Blue - Mama ft. William Singe (Official Video)
Alan Walker & Ruben – Heading Home (Official Music Video)
2 Chainz & Wiz Khalifa - We Own It (Fast & Furious / Official Lyric Video)
(Star Wars Ships) Knights of Ren FLAGSHIP Breakdown "Night Buzzard"
K.S Makhan - Same Size - Latest Punjabi Song
Alan Walker & Ruben – Heading Home (Official Music Video)
Niall Horan - This Town (Lyrics)
Alan Walker - Diamond Heart feat. Sophia Somajo (Lyric Video)
NIve - Tired | Official Animated Video
The Shelters - Rebel Heart [Official Music Video]
Porter Robinson & Madeon - Shelter (Official Video) (Short Film with A-1 Pictures & Crunchyroll)
One Direction - Night Changes
Humraah Song | Malang | Aditya R K, Disha P Anil K Kunal K | Sachet T | Mohit S | Fusion P |Kunaal V
One Direction - History (Official 4K Video)
The Weeknd - Save Your Tears (Official Music Video)
LYRICAL: Bekhayali | Kabir Singh | Shahid K,Kiara A |Sandeep Reddy Vanga | Sachet-Parampara | Irshad
Jonas Blue – Mama (Lyrics) 🎵 ft. William Singe
Katy Perry - Rise (Official)
One Direction - Night Changes
Zedd - Beautiful Now ft. Jon Bellion (Official Music Video)
Jay Sean - Down ft. Lil Wayne (Official Music Video)
Frenship & Emily Warren - Capsize (Official Lyric Video)
Deep Purple - Perfect Strangers (Official Video)
Kygo - Stranger Things ft. OneRepublic (Official Video)
MC Lars - Download This Song [Official Music Video]
Zedd - Beautiful Now ft. Jon Bellion (Official Music Video)
TAKE ME BACK - Official Music Video
twenty one pilots: Stressed Out [OFFICIAL VIDEO]
Alan Walker - The Spectre
Zedd - Beautiful Now ft. Jon Bellion (Official Audio)
Dua Lipa - Swan Song (From Alita: Battle Angel) [Official Music Video]
Felix Jaehn - Ain't Nobody (Loves Me Better) (Official Video) ft. Jasmine Thompson
Hum Hain Rahi Pyar Ke (1993) | Aamir Khan | Juhi Chawla | 90's Superhit Bollywood Songs
Baharon Phool Barsao | Lyrical Song | Suraj 1966 | Mohammed Rafi | Rajendra Kumar, Vyjayanthimala
Song:  All I Have to Do Is GIS
Rita Ora - Your Song (Official Video)
AGE 19 : Jass Manak Ft. DIVINE (Official Song) Deep Jandu | GK.DIGITAL | Geet MP3
Liam Payne - Strip That Down (Official Video) ft. Quavo
Melanie Fiona 4AM Official Lyrics
Zedd, Alessia Cara - Stay (Lyric Video)
One Direction - Night Changes
Love Around The World 2020 - original world peace song
Avicii - Wake Me Up (Official Video)
Melanie Fiona 4AM Official Lyrics
2 Chainz & Wiz Khalifa - We Own It (Fast & Furious / Official Lyric Video)
Jonas Blue, Liam Payne, Lennon Stella - Polaroid (Official Video)
Lost Frequencies & Mathieu Koss - Don't Leave Me Now
One Direction - Night Changes
Bebe Rexha - Meant to Be (feat. Florida Georgia Line) [Official Music Video]
Liam Payne - Strip That Down (Official Video) ft. Quavo
NF - When I Grow Up
Dimitri Vegas & Like Mike ft. Wiz Khalifa - When I Grow Up (Official Music Video)
Jonas Blue - Mama ft. William Singe (Official Video)
Mike Posner - I Took A Pill In Ibiza (Original)
Shawn Mendes - Stitches (Official Video)

tests.py
import tkinter as tk

root = tk.Tk()
pic = tk.PhotoImage(file=r"button.png")
b1 = tk.Button(root, image=pic)
b1.grid(row=1, column=0)

root.mainloop()


downloader.py
import pafy
from googlesearch import search
import tkinter as tk
import sys

wd = tk.Tk()
wd.title('Youtube video downloader')


def button_c(i):
    # global videoideo
    print(i)
    print("downloading.....")
    i.download(quiet=True, callback=mycb)
    print("download completed")


def mycb(total, recvd, ratio, rate, eta):
    sys.stdout.flush()
    total = int(total) / 1000000
    recvd = int(recvd) / 1000000
    total = "{:.2f}".format(total)
    recvd = "{:.2f}".format(recvd)
    rate = "{:.2f}".format(rate)
    time_remaining = 'sec'
    if int(eta) >= 60:
        eta = int(eta) / 60
        time_remaining = 'min'
        eta = "{:.1f}".format(eta)
    else:
        time_remaining = 'sec'

    sys.stdout.write(f"\rtotal = {total}")
    sys.stdout.write(f"\rtotal : {total}MB  recived : {recvd}MB  ETA : {eta}{time_remaining}  speed : {rate}kbps")


song = input("enter the name of the song:-> ")
flag = True
name = song + " official music video youtube"
query = ''
name = name.split()
for a in name:
    query = query + ' ' + a
for urls in search(query, tld="co.in", num=10, stop=1, pause=2):
    if urls.__contains__("www.youtube.com"):
        print(urls)
        break

video = pafy.new(urls)
tk.Label(wd, text=f"{video.title}").grid(row=0, column=2, columnspan=2)
n = 0
ro = 2
for i in video.streams:
    if n > 5:
        ro += 1
        n = 0
    bi = tk.Button(wd, text=i, command=lambda i=i: button_c(i)).grid(row=ro, column=n)
    n += 1
best = video.getbest()
# getbestnl=video.getbest
b2 = tk.Button(wd, text=best, command=lambda best=best: button_c(best)).grid(row=ro, column=n)
n += 1
for i in video.audiostreams:
    if n > 5:
        ro += 1
        n = 0
    bi = tk.Button(wd, text=i, command=lambda i=i: button_c(i)).grid(row=ro, column=n)
    n += 1
wd.mainloop()


music_player.py
import webbrowser
import random
import pafy
from googlesearch import search

print("""
enter'@ny'(any) for random song
'@dd' to add a new song in the library 
'@list' to see the playlist or
'@his' to see the history of played song 
'@down' to download a song
'/q' to exit
simply the name of the song you want to play ↓
""")


def playsong(name):  # give name of the song to play it
    flag = True
    name = name + " song (official video) youtube"
    query = ''
    name = name.split()
    for a in name:
        query = query + ' ' + a
    print(query)
    for urls in search(query, tld="co.in", num=10, stop=1, pause=2):
        if urls.__contains__("www.youtube.com"):
            print(urls)
            break
    webbrowser.open_new_tab(urls)
    videod = pafy.new(urls)
    print("playing -> " + videod.title)
    ab = open('history.txt', 'a+')
    abd = ab.readlines()
    for i in abd:
        if videod.title == i:
            flag = False
        if flag:
            ab.write(videod.title)
            flag = False
        ab.close()


while True:

    a = " "
    search_keyword = input("enter your command:->")
    if search_keyword.strip() == "@ny":

        with open("ex.txt", "r") as f:
            lines = f.readlines()
            line = random.choice(lines)
            print(line)
            playsong(line)

    elif search_keyword.strip() == "@dd":
        with open("ex.txt", "a") as f:
            name = input("name of the song-> ")
            name = name + " official music video"
            query = ''
            name = name.split()
            for a in name:
                query = query + ' ' + a
            for urls in search(query, tld="co.in", num=10, stop=10, pause=2):
                if urls.__contains__("www.youtube.com"):
                    print(urls)
                    break
            videod = pafy.new(urls)
            f.write('\n' + videod.title)
            f.close()
            print("song added to the library -> " + videod.title)
            print("would you like to play it now(y/n)")
            if input(":-> ").lower() == 'y':
                print("playing->" + videod.title)
                webbrowser.open_new_tab(urls)

    elif search_keyword.strip() == "@list":
        with open("ex.txt", "r") as f:
            liness = f.readlines()
        datta = []
        i = 1
        for ab in liness:
            data = f'{i}->{ab}'
            datta.append(data)
            i += 1
            print(data)
        print("""
-> enter the number assigned to the song you wanna play or
-> if you don't want to play any from this simply write the name of the song
-> enter '@del' to delete a song
-> or press escape to go back
            """)
        num = input(":-> ")
        if num.isdigit():
            name = ' '
            for i in datta:
                i = i.split('->')
                if i[0] == num:
                    name = name.join(i[1:])
                    playsong(name)
        elif num == "@del":
            de = input('enter the number of the song you wanna delete-> ')
            dele = ''
            if de.isdigit():
                new = open("ex.txt", "w")
                for i in datta:
                    i = i.split('->')
                    if i[0] == de:
                        dele = dele.join(i[1:])
                for line in liness:
                    if line == dele:
                        print(f'deleting the song ->{dele}')
                    else:
                        new.write(line)
                new.close()
        else:
            playsong(num)

    elif search_keyword.lower().strip() == '@his':
        with open('history.txt', 'r') as s:
            print(s.read())
            # print("****press 'n' to play recent song***")
            hist_command = input("->enter 'del' to delete history or just hit enter:         ")
            if hist_command.lower().strip() == 'del':
                hist_text = open("history.txt", 'w')
                hist_text.close()
                print("***history deleted***")
    elif search_keyword.lower().strip() == '@down':
        import downloader
    elif search_keyword.lower().strip() == "/q":
        quit()
    else:
        playsong(search_keyword)


musicPlayer_tkinter_.py
import tkinter as tk
import webbrowser
from functools import partial
from googlesearch import search
import pafy
from tkinter import ttk
from tkinter import messagebox
import random
import sys

try:

    root = tk.Tk()
    root.geometry('275x107')
    # root.
    root.title('Music Player')

    bg_image = tk.PhotoImage(file=r"name.png")
    # create a main frame
    main_frame = tk.Frame(root)
    main_frame.pack(fill=tk.BOTH, expand=1)

    # create a canvas
    my_canvas = tk.Canvas(main_frame)
    my_canvas.pack(side=tk.LEFT, fill=tk.BOTH, expand=1)

    # add a scrollbar to the canvas
    my_scrollbar = ttk.Scrollbar(main_frame, orient=tk.VERTICAL, command=my_canvas.yview)
    my_scrollbar.pack(side=tk.RIGHT, fill=tk.Y)

    # connfigure  the canvas
    my_canvas.configure(yscrollcommand=my_scrollbar.set)
    my_canvas.bind('<Configure>', lambda e: my_canvas.configure(scrollregion=my_canvas.bbox('all')))

    # create ANOTHER frame inside canvas
    second_frame = tk.Frame(my_canvas)

    # add that new frame to a window in the canvas
    my_canvas.create_window((0, 0), window=second_frame, anchor="nw")


    def _on_mouse_wheel(event):
        my_canvas.yview_scroll(-1 * int((event.delta / 120)), "units")


    def buttonPress():
        a = e1.get()
        e1.delete(-1)
        Playsong(a)


    def callback(event):
        a = e1.get()
        e1.delete(0, 100)
        Playsong(a)


    def clear_text(self):
        self.e1.delete(0, 100)


    def Playsong(name):
        flag = True
        name = name + " song (official video) youtube"
        query = ''
        name = name.split()
        for a in name:
            query = query + ' ' + a
        print(query)
        for urls in search(query, tld="co.in", num=10, stop=1, pause=2):
            if urls.__contains__("www.youtube.com"):
                print(urls)
                break
        videod = pafy.new(urls)
        playy = tk.messagebox.askokcancel(title=None, message=f'play:- "{videod.title} "?')
        if playy:
            webbrowser.open_new_tab(urls)
            ab = open('history.txt', 'a+')
            abd = ab.readlines()
            for i in abd:
                if videod.title == i:
                    flag = False
                if flag:
                    ab.write(videod.title)
                    flag = False
                ab.close()


    def sonf(name):
        webbrowser.open_new_tab(name)


    def song_name(search_word):
        flag = True
        name = search_word + " song (official video) youtube"
        for urls in search(name, tld="co.in", num=10, stop=1, pause=2):
            if urls.__contains__("www.youtube.com"):
                print(urls)
                break
        videod = pafy.new(urls)
        addd = tk.messagebox.askokcancel(title=None, message=f'add:- "{videod.title}" ?')
        if addd:
            with open('ex.txt', 'a') as f:
                f.write(f'\n{videod.title} |-><-| {urls}')
            playy = tk.messagebox.askokcancel(title=None, message=f'play:- "{videod.title}" ?')
            if playy:
                webbrowser.open_new_tab(urls)


    def add_song():

        ent = tk.Entry(second_frame)
        ent.grid(row=1, column=4)

        # searchsong = partial(song_name, e2)
        def songg():
            search_word = ent.get()
            song_name(search_word)

        bb4 = tk.Button(second_frame, text='add', command=songg).grid(row=1, column=5)
        root.geometry('900x400')


    def deel(ab):

        with open("ex.txt", "r") as f:
            liness = f.readlines()
        with open("ex.txt", "w") as f:
            for i in liness:
                if i != ab:
                    f.write(i)
                else:
                    tk.messagebox.showinfo(title='song deleted', message=f'{ab} \n deleted from the playlist')


    def editt():
        with open("ex.txt", "r") as f:
            liness = f.readlines()
        i = 2
        col = 3
        for ab in liness:
            i += 1
            delete = partial(deel, ab)
            buttonn = tk.Button(second_frame, text=f'<delete> {ab}', command=delete)
            buttonn.grid(row=i, column=col, columnspan=4)


    def play_list():
        butt = tk.PhotoImage(file=r"button.png")
        # butt = butt.subsample(1, 1)
        root.geometry()
        # root.geometry('800x400')
        with open("ex.txt", "r") as f:
            liness = f.readlines()
        i = 2
        col = 3
        add = tk.Button(second_frame, text="add a song", command=add_song)
        add.grid(row=1, column=3)
        # edit = tk.Button(second_frame, image=butt)
        edit = tk.Button(second_frame, text='edit', command=editt)
        edit.grid(row=2, column=3, rowspan=2)
        for a in liness:
            ab = a.split("|-><-|")
            i += 1
            url = str(ab[1]).strip()
            playy = partial(sonf, url)
            buttonn = tk.Button(second_frame, text=ab[0], command=playy)
            buttonn.grid(row=i, column=col, columnspan=4)


    def downloader():
        downloading_window = tk.Toplevel(root)
        entry1 = tk.Entry(downloading_window)
        entry1.grid(row=0, column=4)

        def button_c(i):
            print(i)
            print("downloading.....")
            i.download(quiet=False, callback=mycb)
            print("download completed")

        def mycb(total, recvd, ratio, rate, eta):

            # global n, ro
            sys.stdout.flush()
            total = int(total) / 1000000
            recvd = int(recvd) / 1000000
            total = "{:.2f}".format(total)
            recvd = "{:.2f}".format(recvd)
            rate = "{:.2f}".format(rate)
            time_remaining = 'sec'
            if int(eta) >= 60:
                eta = int(eta) / 60
                time_remaining = 'min'
                eta = "{:.1f}".format(eta)
            else:
                time_remaining = 'sec'

            sys.stdout.write(f"\rtotal = {total}")
            sys.stdout.write(
                f"\rtotal : {total}MB  recived : {recvd}MB  ETA : {eta}{time_remaining}  speed : {rate}kbps")

        def get_song(search_word):
            print(search_word)
            name = search_word + " song (official video) youtube"
            for urls in search(name, tld="co.in", num=10, stop=1, pause=2):
                if urls.__contains__("www.youtube.com"):
                    print(urls)
                    break
            video = pafy.new(urls)
            tk.Label(downloading_window, text=f"{video.title}").grid(row=1, column=2, columnspan=2)
            n = 1
            ro = 3
            for i in video.streams:
                if n > 5:
                    ro += 1
                    n = 0
                bi = tk.Button(downloading_window, text=i, command=lambda i=i: button_c(i)).grid(row=ro, column=n)
                n += 1
            n = 0
            ro += 1

            for i in video.audiostreams:
                if n > 5:
                    ro += 1
                    n = 0
                bi = tk.Button(downloading_window, text=i, command=lambda i=i: button_c(i)).grid(row=ro, column=n)
                n += 1

        def songsss():
            downloading_name = entry1.get()
            get_song(downloading_name)

        bb4 = tk.Button(downloading_window, text='download', command=songsss)
        bb4.grid(row=0, column=5)
        # entry1.insert(0, 'enter the name of the song')
        tk.Label(downloading_window, text='enter the name of the song').grid(row=0, column=1, columnspan=3)


    def randomm():
        with open("ex.txt", "r") as f:
            lines = f.readlines()
            line = random.choice(lines)
            linee = line.split('|-><-|')
            ask = tk.messagebox.askyesnocancel(title='random song', message=f'would you like to play: {linee[0]} ?')
            if ask:
                link = str(linee[1].strip())
                webbrowser.open_new_tab(link)


    label1 = tk.Label(second_frame, image=bg_image)
    label1.place(x=0, y=0)
    ab = tk.Label(second_frame, text=' ').grid(row=4, column=1)
    bb1 = tk.Button(second_frame, text='play list', command=play_list)
    bb1.grid(row=1, column=0)
    tk.Label(second_frame, text="enter the name of the song or").grid(row=3, column=0, columnspan=2)
    e1 = tk.Entry(second_frame)
    e1.grid(row=4, column=0, columnspan=2)
    b1 = tk.Button(second_frame, text="play", command=buttonPress)
    b1.grid(row=5, column=0, columnspan=2)
    bb2 = tk.Button(second_frame, text='download', command=downloader)
    bb3 = tk.Button(second_frame, text='random', command=randomm)
    bb2.grid(row=1, column=1)
    bb3.grid(row=1, column=2)

    root.bind("<MouseWheel>", _on_mouse_wheel)

    root.bind('<Return>', callback)

    root.mainloop()


except:
    tk.messagebox.showerror(title='error', message='check your internet connection')
